src/quic/frame_reader.d

```diff
@@ -0,0 +1,98 @@
+module quic.frame_reader;
+
+import quic.frame;
+import quic.packet;
+import quic.attributes;
+
+struct QuicReader(FrameType)
+{
+    import quic.decode : decodeVarInt;
+
+    ubyte[] networkStream;
+    ulong bufIndex;
+
+    this(ubyte[] networkStream, ulong bufIndex)
+    {
+        this.networkStream = networkStream; 
+        this.bufIndex = bufIndex;
+    }
+
+    auto opDispatch(string name)() {
+            alias FieldType = typeof(mixin(FrameType.stringof, '.', name));
+
+            FieldType value;
+
+            static if (is(FieldType == VarInt))
+            {
+                decodeVarInt(value, networkStream, bufIndex);
+                return value;
+            }
+
+            else static if (is(FieldType == AckRange))
+            {
+                decodeVarInt(value.gap, networkStream, bufIndex);
+                decodeVarInt(value.rangeLength, networkStream, bufIndex);
+                return value;
+            }
+
+            else static if ((is(FieldType == EcnCount)))
+            {
+                decodeVarInt(value.ect0Count, networkStream, bufIndex);
+                decodeVarInt(value.ect1Count, networkStream, bufIndex);
+                decodeVarInt(value.ecn_ceCount, networkStream, bufIndex);
+                return value;
+            }
+
+            else static if (is(FieldType == ubyte[]))
+            {
+                alias attributes = __traits(getAttributes, mixin(FrameType.stringof, '.', name)); 
+                static if(is(attributes == VarIntLength))
+                {
+                    VarInt len;
+                    decodeVarInt(len, networkStream, bufIndex);
+                    bufIndex += len; 
+                    return networkStream[bufIndex-len..bufIndex];
+                }
+
+                static if(hasFixedLengh!(attributes[0]))
+                {
+                    auto lenOfLength = getFixedLength!(attributes[0]);
+                    auto len = readBigEndianField(networkStream, bufIndex, len);
+                    bufIndex += len;
+                    return networkStream[bufIndex-len..bufIndex];
+                }
+            }
+
+            else static if(hasEstablishedLength!(attributes[0]))
+            {
+                    auto len = getEstablishedLength!(attributes[0]);
+                    bufIndex += len;
+                    return networkStream[bufIndex-len..bufIndex];
+            }
+        }
```

Comment:
you might also consider a `else static assert(0)` to indicate the type is not supported